### PR TITLE
add gprof support

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
Support gprof function for analyzing code hot spots and function call relationships

usage:
build:
enable CONFIG_SYSTEM_GPROF

nuttx:
nsh> gprof start
nsh> gprof stop
nsh> gprof dump /tmp/gmon.out

ubuntu: 
adb pull /tmp/gmon.out ./gmon.out
arm-none-eabi-gprof ./nuttx/nuttx gmon.out -b 

The saved file format complies with the standard format. 
For specific instructions on command usage, refer to the following link：
https://ftp.gnu.org/old-gnu/Manuals/gprof-2.9.1/html_mono/gprof.html

## Impact
none

## Testing
stm32

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



The PR Summary is missing crucial information:

* **Why is this change necessary?**  Is there a specific performance problem being addressed? 
* **What functional part of the code is being changed?** Mention the specific files/modules modified.
* **How does the change exactly work?** Provide more detail than just "Support gprof". Explain how gprof is integrated, any data structures added, etc.

**Impact Section:**

* **Is new feature added?**  - This seems like a YES.
* **Impact on build?** -  You mention enabling a config option, so this is a YES. Describe what changes in the build process when the option is enabled.
* **Impact on documentation?** -  YES, you'll need to document how to use this new gprof feature.

**Testing Section:**

* **Build Host(s):** Be specific about your build host OS and compiler version.
* **Target(s):**  "stm32" is not specific enough. Provide the exact board and configuration.
* **Testing logs:**  You need to provide actual logs before and after the change to demonstrate the functionality and any performance improvements.

**In short, the PR currently lacks the necessary details and evidence to be considered complete.** 

To improve it, address the missing points mentioned above and provide concrete examples in the Summary, Impact, and Testing sections.

File: include/unistd.h [hunk 1]
```diff
@@ -466,6 +466,9 @@ int     getentropy(FAR void *buffer, size_t length);
 void    sync(void);
 int     syncfs(int fd);
 
+int     profil(FAR unsigned short *buf, size_t bufsiz,
```
Comment (acassis) [inline include/unistd.h:469]: Why to use profil instead of profile?

File: sched/Kconfig [hunk 2]
```diff
@@ -1162,6 +1162,13 @@ config SCHED_CPULOAD_TIMECONSTANT
 		tick count exceeds this time constant.  This time constant is in
 		units of seconds.
 
+config SCHED_PROFIL_TICKSPERSEC
```
Comment (acassis) [inline sched/Kconfig:1165]: Same here, PROFILE instead of PROFIL

File: sched/sched/Make.defs [hunk 3]
```diff
@@ -20,7 +20,7 @@
 #
 ############################################################################
 
-CSRCS += sched_getfiles.c
+CSRCS += sched_getfiles.c sched_profil.c
```
Comment (acassis) [inline sched/sched/Make.defs:23]: Ditto

File: sched/sched/sched_profil.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+ * Private Data
+ ****************************************************************************/
+
+static struct profinfo_s g_prof;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void profil_timer_handler(wdparm_t arg)
```
Comment (acassis) [inline sched/sched/sched_profil.c]: Ditto. Please include the function description

File: sched/sched/sched_profil.c [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+ *   offset - The lowest address to be sampled
+ *   scale  - Multiply address by scale / 65536
+ *
+ * Returned Value:
+ *   Zero (OK) if successful. Otherwise, ERROR is returned and
+ *   errno is set to indicate the error.
+ *
+ ****************************************************************************/
+
+int profil(FAR unsigned short *buf, size_t bufsiz,
```
Comment (acassis) [inline sched/sched/sched_profil.c:128]: Ditto

**Comment by acassis** [comment]
@Gary-Hobson Please include an initial Documentation/. Basically include the usage examples you put in the Summary and a brief explanation of GNU Profile, explaining it could be used to detect code bottlenecks and help to optimize the code.

File: sched/Kconfig (hunk 2 above)
Comment (xiaoxiang781216) [reply sched/Kconfig:1165]: ditto

**Review by cederom — CHANGES_REQUESTED** [review]
Thank you @Gary-Hobson :-)

As requested in https://github.com/apache/nuttx/pull/13999#issuecomment-2402259024 please also update documentation when introducing new features.

Also please fill in the Impact and Testing sections in PR we will not accept PR with those sections invalid / empty.

File: sched/Kconfig (hunk 2 above)
Comment (acassis) [reply sched/Kconfig:1165]: Yes, I think here in the menuconfig PROFILE name makes more sense

**Comment by xiaoxiang781216** [comment]
please fix the conflict

File: sched/Kconfig (hunk 2 above)
Comment (Gary-Hobson) [reply sched/Kconfig]: Done

**Comment by xiaoxiang781216** [comment]
@Gary-Hobson please fix ci error.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]
